src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,471 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// The exported function names need to be unique (can't be disambiguated based on signature), hence
+// we introduce suffix letters to indicate the general patterns used.
+// * U suffix means unaligned and unpadded.
+// * S suffix means sparse (unaligned) vector.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    internal static class SseIntrinsics
+    {
+        private static unsafe Vector128<float> Load1(float* src, int* idx)
+        {
+            return Sse.SetScalarVector128(src[idx[0]]);
+        }
+
+        private static unsafe Vector128<float> Load4(float* src, int* idx)
+        {
+            return Sse.SetVector128(src[idx[3]], src[idx[2]], src[idx[1]], src[idx[0]]);
+        }
+
+        private static Vector128<float> Rotate(Vector128<float> x)
+        {
+            // The control byte shuffles the four 32-bit floats of x: ABCD -> BCDA.
+            return Sse.Shuffle(x, x, 0x39);
+        }
+
+        private static Vector128<float> RotateReverse(Vector128<float> x)
+        {
+            // The control byte shuffles the four 32-bit floats of x: ABCD -> DABC.
+            return Sse.Shuffle(x, x, 0x93);
+        }
+
+        private static unsafe void Store4(Vector128<float> x, float* dst, int* idx)
+        {
+            Sse.StoreScalar(dst + idx[0], x);
+            x = Rotate(x);
+            Sse.StoreScalar(dst + idx[1], x);
+            x = Rotate(x);
+            Sse.StoreScalar(dst + idx[2], x);
+            x = Rotate(x);
+            Sse.StoreScalar(dst + idx[3], x);
+        }
+
+        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
+        private static void ToVectorSum(ref Vector128<float> vector)
+        {
+            if (Sse3.IsSupported)
+            {
+                vector = Sse3.HorizontalAdd(vector, vector);
+                vector = Sse3.HorizontalAdd(vector, vector);
+            }
+            else
```

Comment:
@tannergooding: (Responding to your previous PR comment) Factoring the `VectorSum` function out of the intrinsics has worked quite well, but my old implementation of `VectorSum` (which outputs a `Vector128<float>`) seemed to cause some perf issues (doubled the running time of 4 intrinsics that call `VectorSum`).  I changed it to `ToVectorSum`, which changes `vector` in place, but the same perf issue persisted.  Then, I (thanks to Santi's help) tried the `AggressiveInlining` attribute above to make the function inline, which has improved the perf results back to normal.  Curious about how it sounds to you, and if there is any better way to do it.

Side question: Should I also make the other private functions - `Load1`, `Load4`, `Rotate`, `RotateReverse`, and `Store4` - inline with the `AggressiveInlining` attribute to improve perf?

Reply:
We'll probably want to change `private static void ToVectorSum(ref Vector128<float> vector)` to `private static Vector128<T> ToVectorSum(in Vector128<float> vector)`, as `Vector128<T>` was made into a `readonly struct`, but it otherwise looks good.

> Should I also make the other private functions - Load1, Load4, Rotate, RotateReverse, and Store4 - inline with the AggressiveInlining attribute to improve perf

I would think they are already being inlined (except for Store4), if that isn't the case, then the methods are probably small enough that inlining would help. I would recommend benchmarking and looking at the codegen for both having them explicitly inlined and not.

Reply:
It does not hurt to put AggressiveInlining attribute on these short and internal functions. 

For your old code, even though VectorSum is inlined, the codegen is not optimized due to register spill and reload. It seems JIT has optimization opportunity over there. Do you mind opening an issue to discuss about it on CoreCLR github repo?

Reply:
Thanks for the input!  Adding `AggressiveInlining` attributes for those short functions to mimic the behavior of the existing native code.

Reply:
I'll pick up logging a bug for CoreCLR. We should get a minimal repro and ensure it isn't already tracked by one of the "first class struct" issues.